firedrake/function.py

```diff
@@ -698,6 +699,93 @@ def __str__(self):
         return "domain %s does not contain point %s" % (self.domain, self.point)
 
 
+class PointEvaluator:
+    r"""Convenience class for evaluating a :class:`Function` at a set of points."""
+
+    def __init__(self, mesh: MeshGeometry, points: np.ndarray | list, tolerance: float | None = None,
+                 missing_points_behaviour: str | None = "error", redundant: bool = True) -> None:
+        r"""
+        Parameters
+        ----------
+        mesh : :class:`MeshGeometry`
+            The :class:`Mesh` on which the :class:`Function` is defined.
+        points : np.ndarray | list
+            Array of points to evaluate the function at.
+        tolerance : float | None
+            Tolerance to use when checking if a point is in a cell. Default is the ``tolerance`` of the :class:`Mesh`.
+        missing_points_behaviour : str | None
+            Behaviour when a point is not found in the mesh. Options are:
+            "error": raise a :class:`VertexOnlyMeshMissingPointsError` if a point is not found in the mesh.
+            "warn": warn if a point is not found in the mesh, but continue.
+            None: ignore points not found in the mesh.
+        redundant : bool
+            If True, only the points on rank 0 are evaluated, and the result is broadcast to all ranks.
+            If False, each rank evaluates the points it has been given.
+            Default is True.
+        """
+        from firedrake.mesh import VertexOnlyMesh
+        self.mesh = mesh
+        self.points = np.asarray(points, dtype=utils.ScalarType)
+        self.redundant = redundant
+        self.vom = VertexOnlyMesh(
+            mesh, points, missing_points_behaviour=missing_points_behaviour,
+            redundant=redundant, tolerance=tolerance
+        )
+
+    def evaluate(self, function: Function) -> np.ndarray | list[np.ndarray]:
+        r"""Evaluate the given :class:`Function` at the points provided to this
+        :class:`PointEvaluator`. Points that were not found in the mesh will be
+        evaluated to np.nan.
+
+        Parameters
+        ----------
+        function : :class:`Function`
+            The :class:`Function` to evaluate.
+
+        Returns
+        -------
+        np.ndarray | list[np.ndarray]
```

Comment:
Personally I prefer to return a tuple here because we want the result to be immutable.